kong/plugins/key-auth/handler.lua

```diff
@@ -148,6 +158,54 @@ local function do_authentication(conf)
     return nil, { status = 401, message = "No API key found in request" }
   end
 
+   -- search for signatire in headers & querystring
+   if conf.validate_signature then
+    for i = 1, #conf.signature_names do
+      local name = conf.signature_names[i]
+      local v = headers[name]
+      if not v then
+        -- search in querystring
+        v = query[name]
+      end
+
+      -- search the body, if we asked to
+      if not v and conf.signature_in_body then
+        v = body[name]
+      end
+
+      if type(v) == "string" then
+        signature = v
+
+        if conf.hide_credentials then
+          query[name] = nil
+          kong.service.request.set_query(query)
+          kong.service.request.clear_header(name)
+
+          if conf.key_in_body then
+            body[name] = nil
+            kong.service.request.set_body(body)
+          end
+        end
+
+        break
+
+      elseif type(v) == "table" then
+        -- duplicate API key
+        return nil, { status = 401, message = "Duplicate signature found" }
+      end
+    end
+
+    -- this request is missing a signature, HTTP 401
+    if not signature or signature == "" then
+      kong.response.set_header("WWW-Authenticate", _realm)
+    return nil, { status = 401, message = "No signature found in request" }
+  end
+  end
```

Comment:
This duplicates almost all of the above code to find the API key. Could you please try to factorise it in its own utility function instead, so we avoid duplicating such a large amount of (tricky) code? Thanks.